docs/shaders.rst

```diff
@@ -0,0 +1,325 @@
+                ...
+                var out = get_fragment_output(varyings.position.z, color);
+                return out;
+            }
+            """
+
+Picking
+-------
+
+The output struct in the previous section also has a ``pick`` field that can
+be set with picking info. It is an ``u64`` into which we can pack as many fields
+as needed, using the ``pick_pack()`` function. The material needs to
+implement a corresponding ``_wgpu_get_pick_info()`` method
+to unpack the picking info. See e.g. the picking of a mesh:
+
+.. code-block:: python
+
+        def code_fragment(self):
+            return """
+            @stage(fragment)
+            fn fs_main(varyings: Varyings) -> FragmentOutput {
+                ...
+                var out = get_fragment_output(varyings.position.z, color);
+
+                // The write_pick templating variable is builtin too
+                $$ if write_pick
+                // 20 + 26 + 6 + 6 + 6 = 64
+                out.pick = (
+                    pick_pack(varyings.pick_id, 20) +
+                    pick_pack(varyings.pick_idx, 26) +
+                    pick_pack(u32(varyings.pick_coords.x * 64.0), 6) +
+                    pick_pack(u32(varyings.pick_coords.y * 64.0), 6) +
+                    pick_pack(u32(varyings.pick_coords.z * 64.0), 6)
+                );
+                $$ endif
+
+                return out;
+            }
+            """
+
+
+Clipping planes
+---------------
+
+For common features that apply to all/most objects, wgsl convenience functions
+are available. Take clipping planes. One can call ``apply_clipping_planes()`` to
+discard the fragment if it's outside of the clipping planes. Or use
+``check_clipping_planes()`` to get a boolean.
+
+.. code-block:: python
+
+        def code_fragment(self):
+            return """
+            @stage(fragment)
+            fn fs_main(varyings: Varyings) -> FragmentOutput {
+                ...
+
+                apply_clipping_planes(varyings.world_pos);
+                var out = get_fragment_output(varyings.position.z, color);
+                ...
+                return out;
+            }
+            """
+
+
+Colormapping
+------------
+
+Many materials in PyGfx support colormapping. We distinguis between colormaps
+with image input data, and vertex input data (texture coordinates). The number of
+channels of the input data must match the dimension of the colormap (1D, 2D or 3D).
+
+The base shader class has two corresponding helper functions, and there
+is a wgsl helper function.
+
+For images / volumes:
+
+.. code-block:: python
+
+        def get_resources(self, wobjwect, shared):
+            ...
+            extra_bindings = self.define_img_colormap(material.map)
+            ...
+
+        def code_fragment(self):
+            return """
+            @stage(fragment)
+            fn fs_main(varyings: Varyings) -> FragmentOutput {
+                ...
+                let color = sample_colormap(img_valu);
```

Comment:
where does `img_valu` come from? also, should it be `img_value`?

Reply:
From sampling the image texture. Added an extra line above to clarify this.